internal/driver/onvifdiscovery.go

```diff
@@ -0,0 +1,235 @@
+		discovered = sdkModel.DiscoveredDevice{
+			Name:        endpointRefAddr,
+			Protocols:   dev.Protocols,
+			Description: "Auto discovered Onvif camera",
+			Labels:      []string{"auto-discovery"},
+		}
+		d.lc.Debugf("Discovered unknown camera from the address '%s'", xaddr)
+	} else {
+		dev.Protocols[OnvifProtocol][Manufacturer] = devInfo.Manufacturer
+		dev.Protocols[OnvifProtocol][Model] = devInfo.Model
+		dev.Protocols[OnvifProtocol][FirmwareVersion] = devInfo.FirmwareVersion
+		dev.Protocols[OnvifProtocol][SerialNumber] = devInfo.SerialNumber
+		dev.Protocols[OnvifProtocol][HardwareId] = devInfo.HardwareId
+
+		// Spaces are not allowed in the device name
+		deviceName := fmt.Sprintf("%s-%s-%s",
+			strings.ReplaceAll(devInfo.Manufacturer, " ", "-"),
+			strings.ReplaceAll(devInfo.Model, " ", "-"),
+			endpointRefAddr)
+
+		discovered = sdkModel.DiscoveredDevice{
+			Name:        deviceName,
+			Protocols:   dev.Protocols,
+			Description: fmt.Sprintf("%s %s Camera", devInfo.Manufacturer, devInfo.Model),
+			Labels:      []string{"auto-discovery", devInfo.Manufacturer, devInfo.Model},
+		}
+		d.lc.Debugf("Discovered camera from the address '%s'", xaddr)
+	}
+	return discovered, nil
+}
+
+// mapProbeResults converts a slice of discovered onvif.Device into the generic
+// netscan.ProbeResult.
+func mapProbeResults(host, port string, devices []onvif.Device) (res []netscan.ProbeResult) {
+	for _, dev := range devices {
+		res = append(res, netscan.ProbeResult{
+			Host: host,
+			Port: port,
+			Data: dev,
+		})
+	}
+	return res
+}
+
+// executeRawProbe essentially performs a UDP unicast ws-discovery probe by sending the
+// probe message directly over the connection and listening for any responses. Those
+// responses are then converted into a slice of onvif.Device.
+func executeRawProbe(conn net.Conn, params netscan.Params) ([]onvif.Device, error) {
+	probeSOAP := wsdiscovery.BuildProbeMessage(uuid.Must(uuid.NewV4()).String(), nil, nil,
+		map[string]string{"dn": "http://www.onvif.org/ver10/network/wsdl"})
+
+	addr := conn.RemoteAddr().String()
+	if err := conn.SetDeadline(time.Now().Add(params.Timeout)); err != nil {
+		return nil, errors.Wrapf(err, "%s: failed to set read/write deadline", addr)
+	}
+
+	if _, err := conn.Write([]byte(probeSOAP.String())); err != nil {
+		return nil, errors.Wrap(err, "failed to write probe message")
+	}
+
+	var responses []string
+	buf := make([]byte, bufSize)
+	// keep reading from the PacketConn until the read deadline expires or an error occurs
+	for {
+		n, _, err := (conn.(net.PacketConn)).ReadFrom(buf)
+		if err != nil {
+			// ErrDeadlineExceeded is expected once the read timeout is expired
+			if !errors.Is(err, os.ErrDeadlineExceeded) {
+				params.Logger.Debugf("Unexpected error occurred while reading ws-discovery responses: %s", err.Error())
+			}
+			break
+		}
+		responses = append(responses, string(buf[0:n]))
+	}
+
+	if len(responses) == 0 {
+		params.Logger.Tracef("%s: No Response", addr)
```

Comment:
```suggestion
		params.Logger.Debugf("%s: No Response", addr)
```

Reply:
The reason I chose trace instead of debug is that when using UDP there is no handshake so no response is returned for every single IP address that does not respond, meaning you can have up to 254 logs in a standard `/24` subnet.

The more interesting debug logs look like so: `msg="read udp 192.168.0.190:41133->192.168.0.1:3702: recvfrom: connection refused"` meaning the host is online but refusing the connection.